lib/widgets/sing_in_buttons.dart

```diff
@@ -0,0 +1,180 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io' show Platform;
+import 'dart:math';
+import 'package:crypto/crypto.dart';
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:google_sign_in/google_sign_in.dart';
+import 'package:sign_in_with_apple/sign_in_with_apple.dart' as apple;
+
+class SignInButtons extends StatefulWidget {
+  const SignInButtons({
+    this.onSignOut,
+    this.onSingIn,
+    Key? key,
+  }) : super(key: key);
+
+  final FutureOr<void> Function()? onSingIn;
+  final FutureOr<void> Function()? onSignOut;
+
+  @override
+  _SignInButtonsState createState() => _SignInButtonsState();
+}
+
+class _SignInButtonsState extends State<SignInButtons> {
+  var loading = false;
+
+  Future<void> signIn(
+    Future<AuthCredential?> Function() getCredentials,
+  ) async {
+    await signOut();
+    setState(() {
+      loading = true;
+    });
+    final credentials = await getCredentials();
+    if (credentials != null) {
+      await FirebaseAuth.instance.signInWithCredential(credentials);
+      await widget.onSingIn?.call();
+    }
+    setState(() {
+      loading = false;
+    });
+  }
+
+  Future<void> signOut() async {
+    setState(() {
+      loading = true;
+    });
+    final provider =
+        FirebaseAuth.instance.currentUser?.providerData.first.providerId;
+    if (provider == null) return;
+
+    await FirebaseAuth.instance.signOut();
+    switch (provider) {
+      case 'google.com':
+        await GoogleSignIn().signOut();
+        break;
+      case 'apple.com':
+        // TODO await appleSignOut;
+        break;
+      default:
+    }
+    await widget.onSignOut?.call();
+    setState(() {
+      loading = false;
+    });
+  }
+
+  Future<AuthCredential?> getGoogleCredentials() async {
+    final user = await GoogleSignIn().signIn();
+    if (user != null) {
+      final auth = await user.authentication;
+      return GoogleAuthProvider.credential(
+        accessToken: auth.accessToken,
+        idToken: auth.idToken,
+      );
+    }
+  }
+
+  Future<AuthCredential?> getAppleCredentials() async {
+    String generateNonce([int length = 32]) {
+      const charset =
+          '0123456789ABCDEFGHIJKLMNOPQRSTUVXYZabcdefghijklmnopqrstuvwxyz-._';
+      final random = Random.secure();
+      return List.generate(
+          length, (_) => charset[random.nextInt(charset.length)]).join();
+    }
+
+    String sha256ofString(String input) {
+      final bytes = utf8.encode(input);
+      final digest = sha256.convert(bytes);
+      return digest.toString();
+    }
+
+    if (kIsWeb) {
```

Comment:
Is there a working way to unify those two platforms? @aaboundegm mentioned the example in their documentation, but I've been told it doesn't work for some reason. If not, that's fine.

Reply:
yeah the IOS part didn't work with the web.